File: mobile-android/app/src/main/java/com/agroerp/domain/engine/FormValidationEngine.kt
package com.agroerp.domain.engine

import com.agroerp.domain.model.FormField
import com.agroerp.domain.model.FormSchema
import com.agroerp.domain.model.GpsPoint
import javax.inject.Inject
import javax.inject.Singleton

class FormValidationException(
    val errors: List<String>,
) : Exception(errors.joinToString("; "))

@Singleton
class FormValidationEngine @Inject constructor(
    private val conditional: ConditionalLogicEngine,
    private val calculated: CalculatedFieldEngine,
) {

    companion object {
        private val LAYOUT_TYPES = setOf(
            "heading", "separator", "html", "markdown", "hyperlink",
            "button", "indicator", "hidden",
        )
        private val GEO_TYPES = setOf("geo", "geo_point", "map")
    }

    fun validate(
        schema: FormSchema,
        rawData: Map<String, Any?>,
        gpsLocation: GpsPoint? = null,
    ): Map<String, Any?> {
        val withCalculated = calculated.resolve(schema.fields, rawData)
        val validated = mutableMapOf<String, Any?>()
        val errors = mutableListOf<String>()

        if (schema.settings.requireGps && gpsLocation == null) {
            errors.add("Se requiere ubicación GPS para este formulario")
        }

        schema.settings.geofence?.let { fence ->
            if (gpsLocation != null) {
                val center = fence["center"] as? Map<*, *>
                val radius = (fence["radiusMeters"] as? Number)?.toDouble()
                if (center != null && radius != null) {
                    val lat = (center["lat"] as? Number)?.toDouble() ?: return@let
                    val lng = (center["lng"] as? Number)?.toDouble() ?: return@let
                    val distance = haversineMeters(gpsLocation.lat, gpsLocation.lng, lat, lng)
                    if (distance > radius) {
                        errors.add("La ubicación GPS está fuera del área permitida")
                    }
                }
            }
        }

        for (field in schema.fields) {
            if (field.type in LAYOUT_TYPES) continue
            if (field.type == "calculated" || field.type == "derived") continue
            if (!conditional.isVisible(field.visibleWhen, withCalculated)) continue

            val required = conditional.isRequired(
                field.required,
                field.requiredWhen,
                withCalculated,
            )
            val value = withCalculated[field.key]
            val hasValue = value != null && value != ""

            if (required && !hasValue) {
                errors.add("El campo \"${field.label}\" es obligatorio")
                continue
            }
            if (!hasValue) continue

            try {
                validated[field.key] = validateField(field, value)
            } catch (e: Exception) {
                errors.add(e.message ?: "Invalid field ${field.key}")
            }
        }

        if (errors.isNotEmpty()) throw FormValidationException(errors)
        return validated
    }

    private fun validateField(field: FormField, value: Any?): Any? = when (field.type) {
        "text", "barcode" -> validateText(field, value)
        "number" -> validateNumber(field, value)
        "boolean" -> validateBoolean(value)
        "checkbox" -> {
            if (field.options.isNotEmpty()) validateMultiSelect(field, value)
            else validateBoolean(value)
        }
        "radio", "select" -> validateSelect(field, value)
        "multi_select" -> validateMultiSelect(field, value)
        "date", "datetime", "time" -> value.toString()
        in GEO_TYPES -> validateGeo(field, value)
        "geo_track" -> validateGeoTrack(value)
        "photo", "video", "audio", "signature", "file" -> validateMediaRef(value)
        "relation" -> value.toString()
        else -> value
    }

    private fun validateText(field: FormField, value: Any?): String {
        val str = value.toString()
        val v = field.validation
        v?.get("minLength")?.let {
            if (str.length < (it as Number).toInt()) {
                throw IllegalArgumentException("Min length ${it.toInt()}")
            }
        }
        v?.get("maxLength")?.let {
            if (str.length > (it as Number).toInt()) {
                throw IllegalArgumentException("Max length ${it.toInt()}")
            }
        }
        v?.get("pattern")?.let {
            if (!Regex(it.toString()).matches(str)) {
                throw IllegalArgumentException("Pattern mismatch")
            }
        }
        return str
    }

    private fun validateNumber(field: FormField, value: Any?): Double {
        val num = (value as? Number)?.toDouble() ?: value.toString().toDoubleOrNull()
            ?: throw IllegalArgumentException("Must be a number")
        field.validation?.get("min")?.let {
            if (num < (it as Number).toDouble()) throw IllegalArgumentException("Below min")
        }
        field.validation?.get("max")?.let {
            if (num > (it as Number).toDouble()) throw IllegalArgumentException("Above max")
        }
        return num
    }

    private fun validateBoolean(value: Any?): Boolean = when (value) {
        is Boolean -> value
        "true" -> true
        "false" -> false
        else -> throw IllegalArgumentException("Must be boolean")
    }

    private fun validateSelect(field: FormField, value: Any?): String {
        val str = value.toString()
        val allowed = field.options.map { it.value }
        if (allowed.isNotEmpty() && str !in allowed) {
            throw IllegalArgumentException("Opción no válida para ${field.label}")
        }
        return str
    }

    private fun validateMultiSelect(field: FormField, value: Any?): List<String> {
        val list = value as? List<*> ?: throw IllegalArgumentException("Debe seleccionar al menos una opción")
        val allowed = field.options.map { it.value }.toSet()
        return list.map { item ->
            val str = item.toString()
            if (allowed.isNotEmpty() && str !in allowed) {
                throw IllegalArgumentException("Opción no válida: $str")
            }
            str
        }
    }

    private fun validateGeo(field: FormField, value: Any?): Map<String, Any?> {
        val map = value as? Map<*, *> ?: throw IllegalArgumentException("Must be geo object")
        val lat = (map["lat"] as? Number)?.toDouble() ?: throw IllegalArgumentException("Invalid lat")
        val lng = (map["lng"] as? Number)?.toDouble() ?: throw IllegalArgumentException("Invalid lng")
        val result = mutableMapOf<String, Any?>("lat" to lat, "lng" to lng)
        val accuracy = (map["accuracy"] as? Number)?.toFloat()
        if (accuracy != null) {
            result["accuracy"] = accuracy
            val maxAcc = (field.validation?.get("maxAccuracyMeters") as? Number)?.toDouble()
            // Advisory only: outdoor GPS often exceeds studio thresholds; still store the value.
            if (maxAcc != null && accuracy > maxAcc) {
                // do not block local submit — server accepts coordinates
            }
        }
        return result
    }

    private fun validateGeoTrack(value: Any?): List<Map<String, Any?>> {
        val list = value as? List<*> ?: throw IllegalArgumentException("Track must be array")
        if (list.isEmpty()) throw IllegalArgumentException("Track cannot be empty")
        return list.map { validateGeo(FormField(key = "t", type = "geo", label = ""), it) }
    }

    private fun validateMediaRef(value: Any?): Any {
        val uuidRegex = Regex(
            """^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$""",
            RegexOption.IGNORE_CASE,
        )
        if (value is List<*>) {
            return value.map {
                val id = it.toString()
                if (!uuidRegex.matches(id)) throw IllegalArgumentException("Invalid media id")
                id
            }
        }
        val id = value.toString()
        if (!uuidRegex.matches(id)) throw IllegalArgumentException("Invalid media id")
        return id
    }

    private fun haversineMeters(lat1: Double, lng1: Double, lat2: Double, lng2: Double): Double {
        val r = 6371000.0
        val dLat = Math.toRadians(lat2 - lat1)
        val dLng = Math.toRadians(lng2 - lng1)
        val a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
            Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
            Math.sin(dLng / 2) * Math.sin(dLng / 2)
        return r * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
    }
}

File: backend/src/core/forms/application/form-validation.engine.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import {
  FormDefinitionSchema,
  FormFieldDefinition,
  FormFieldType,
  FormGeofence,
} from '@agroerp/shared';
import { ConditionalLogicEngine } from './conditional-logic.engine';
import { CalculatedFieldEngine } from './calculated-field.engine';
import { flattenFields, isDataField, isLayoutFieldType, normalizeFieldType } from './field-type.util';

export interface ValidationResult {
  data: Record<string, unknown>;
  fieldResults: { key: string; valid: boolean; error?: string }[];
}

@Injectable()
export class FormValidationEngine {
  constructor(
    private readonly conditional: ConditionalLogicEngine,
    private readonly calculated: CalculatedFieldEngine,
  ) {}

  validate(
    schema: FormDefinitionSchema,
    rawData: Record<string, unknown>,
    options?: {
      gpsLocation?: { lat: number; lng: number; accuracy?: number } | null;
      emitFieldResults?: boolean;
    },
  ): ValidationResult {
    const withCalculated = this.calculated.resolve(schema.fields, rawData);
    const validated: Record<string, unknown> = {};
    const fieldResults: { key: string; valid: boolean; error?: string }[] = [];
    const errors: string[] = [];

    if (schema.settings?.requireGps && !options?.gpsLocation) {
      errors.push('GPS location is required for this form');
    }

    if (schema.settings?.geofence && options?.gpsLocation) {
      const inside = this.isInsideGeofence(
        options.gpsLocation,
        schema.settings.geofence,
      );
      if (!inside) {
        errors.push('GPS location is outside the allowed geofence');
      }
    }

    const allFields = flattenFields(schema.fields);

    for (const field of allFields) {
      if (!isDataField(field)) continue;
      if (field.type === 'calculated' || field.type === 'derived') continue;
      if (isLayoutFieldType(field.type)) continue;

      const visible = this.conditional.isVisible(field.visibleWhen, withCalculated);
      if (!visible) continue;

      const required = this.conditional.isRequired(
        field.required,
        field.requiredWhen,
        withCalculated,
      );

      const value = withCalculated[field.key];
      const hasValue = value !== undefined && value !== null && value !== '';

      if (required && !hasValue) {
        const msg = `Field "${field.key}" is required`;
        errors.push(msg);
        fieldResults.push({ key: field.key, valid: false, error: msg });
        continue;
      }

      if (!hasValue) {
        fieldResults.push({ key: field.key, valid: true });
        continue;
      }

      try {
        validated[field.key] = this.validateField(field, value);
        fieldResults.push({ key: field.key, valid: true });
      } catch (err) {
        const msg = (err as Error).message;
        errors.push(msg);
        fieldResults.push({ key: field.key, valid: false, error: msg });
      }
    }

    if (errors.length > 0) {
      throw new BadRequestException({
        message: 'Form validation failed',
        errors,
        fieldResults: options?.emitFieldResults ? fieldResults : undefined,
      });
    }

    return { data: validated, fieldResults };
  }

  private validateField(field: FormFieldDefinition, value: unknown): unknown {
    const type = normalizeFieldType(field.type);
    switch (type) {
      case 'text':
      case 'barcode':
        return this.validateText(field, value);
      case 'number':
        return this.validateNumber(field, value);
      case 'boolean':
        return this.validateBoolean(value);
      case 'checkbox':
        if (field.options?.length) {
          return this.validateMultiSelect(field, value);
        }
        return this.validateBoolean(value);
      case 'radio':
        return this.validateSelect(field, value);
      case 'date':
      case 'datetime':
        return this.validateDate(value, type === 'datetime');
      case 'select':
        return this.validateSelect(field, value);
      case 'multi_select':
        return this.validateMultiSelect(field, value);
      case 'geo':
        return this.validateGeo(field, value);
      case 'geo_track':
        return this.validateGeoTrack(value);
      case 'photo':
      case 'video':
      case 'audio':
      case 'signature':
      case 'file':
        return this.validateMediaRef(field, value);
      case 'relation':
        return this.validateRelation(field, value);
      case 'repeat_group':
        if (!Array.isArray(value)) {
          throw new Error(`Field "${field.key}" must be an array`);
        }
        return value;
      case 'matrix':
      case 'scale':
      case 'likert':
      case 'rating':
      case 'emoji':
        return value;
      default:
        return this.validateText(field, value);
    }
  }

  private validateText(field: FormFieldDefinition, value: unknown): string {
    if (typeof value !== 'string') {
      throw new Error(`Field "${field.key}" must be a string`);
    }
    const v = field.validation;
    if (v?.minLength && value.length < v.minLength) {
      throw new Error(`Field "${field.key}" min length is ${v.minLength}`);
    }
    if (v?.maxLength && value.length > v.maxLength) {
      throw new Error(`Field "${field.key}" max length is ${v.maxLength}`);
    }
    if (v?.pattern && !new RegExp(v.pattern).test(value)) {
      throw new Error(`Field "${field.key}" does not match pattern`);
    }
    return value;
  }

  private validateNumber(field: FormFieldDefinition, value: unknown): number {
    const num = typeof value === 'number' ? value : Number(value);
    if (Number.isNaN(num)) {
      throw new Error(`Field "${field.key}" must be a number`);
    }
    const v = field.validation;
    if (v?.min !== undefined && num < v.min) {
      throw new Error(`Field "${field.key}" min is ${v.min}`);
    }
    if (v?.max !== undefined && num > v.max) {
      throw new Error(`Field "${field.key}" max is ${v.max}`);
    }
    return num;
  }

  private validateBoolean(value: unknown): boolean {
    if (typeof value === 'boolean') return value;
    if (value === 'true') return true;
    if (value === 'false') return false;
    throw new Error('Value must be a boolean');
  }

  private validateDate(value: unknown, includeTime: boolean): string {
    const date = new Date(value as string);
    if (Number.isNaN(date.getTime())) {
      throw new Error('Value must be a valid date');
    }
    return includeTime ? date.toISOString() : date.toISOString().slice(0, 10);
  }

  private validateSelect(field: FormFieldDefinition, value: unknown): string {
    const str = String(value);
    const allowed = field.options?.map((o) => o.value) ?? field.validation?.enum;
    if (allowed && !allowed.includes(str)) {
      throw new Error(`Field "${field.key}" must be one of: ${allowed.join(', ')}`);
    }
    return str;
  }

  private validateMultiSelect(
    field: FormFieldDefinition,
    value: unknown,
  ): string[] {
    if (!Array.isArray(value)) {
      throw new Error(`Field "${field.key}" must be an array`);
    }
    const allowed = field.options?.map((o) => o.value) ?? field.validation?.enum;
    for (const item of value) {
      const str = String(item);
      if (allowed && !allowed.includes(str)) {
        throw new Error(`Field "${field.key}" contains invalid option: ${str}`);
      }
    }
    return value.map(String);
  }

  private validateGeo(
    field: FormFieldDefinition,
    value: unknown,
  ): { lat: number; lng: number; accuracy?: number } {
    if (
      typeof value !== 'object' ||
      value === null ||
      !('lat' in value) ||
      !('lng' in value)
    ) {
      throw new Error(`Field "${field.key}" must be { lat, lng }`);
    }
    const geo = value as { lat: unknown; lng: unknown; accuracy?: unknown };
    const lat = Number(geo.lat);
    const lng = Number(geo.lng);
    if (Number.isNaN(lat) || Number.isNaN(lng)) {
      throw new Error(`Field "${field.key}" lat/lng must be numbers`);
    }
    if (lat < -90 || lat > 90 || lng < -180 || lng > 180) {
      throw new Error(`Field "${field.key}" coordinates out of range`);
    }
    const result: { lat: number; lng: number; accuracy?: number } = { lat, lng };
    if (geo.accuracy !== undefined) {
      result.accuracy = Number(geo.accuracy);
      // maxAccuracyMeters is advisory for field capture: outdoor GPS often
      // reports 30–100m and rejecting it left Android sync permanently pending.
      const maxAcc = field.validation?.maxAccuracyMeters;
      if (
        maxAcc &&
        Number.isFinite(result.accuracy) &&
        result.accuracy > maxAcc
      ) {
        // keep coordinates; do not fail the submission
      }
    }
    if (field.validation?.requireGps && !result.accuracy) {
      throw new Error(`Field "${field.key}" requires GPS accuracy`);
    }
    return result;
  }

  private validateGeoTrack(value: unknown): { lat: number; lng: number }[] {
    if (!Array.isArray(value) || value.length === 0) {
      throw new Error('Geo track must be a non-empty array of points');
    }
    return value.map((point, i) => {
      try {
        return this.validateGeo(
          { key: `track[${i}]`, type: 'geo', label: '' },
          point,
        );
      } catch {
        throw new Error(`Geo track point ${i} is invalid`);
      }
    });
  }

  private validateMediaRef(
    field: FormFieldDefinition,
    value: unknown,
  ): string | string[] {
    if (field.type === 'photo' || field.type === 'file') {
      if (Array.isArray(value)) {
        return value.map((v) => this.assertUuid(String(v), field.key));
      }
      return this.assertUuid(String(value), field.key);
    }
    return this.assertUuid(String(value), field.key);
  }

  private validateRelation(field: FormFieldDefinition, value: unknown): string {
    return this.assertUuid(String(value), field.key);
  }

  private assertUuid(value: string, key: string): string {
    const uuidRegex =
      /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
    if (!uuidRegex.test(value)) {
      throw new Error(`Field "${key}" must be a valid resource/file id`);
    }
    return value;
  }

  private isInsideGeofence(
    point: { lat: number; lng: number },
    geofence: FormGeofence,
  ): boolean {
    const distance = this.haversineMeters(
      point.lat,
      point.lng,
      geofence.center.lat,
      geofence.center.lng,
    );
    return distance <= geofence.radiusMeters;
  }

  private haversineMeters(
    lat1: number,
    lng1: number,
    lat2: number,
    lng2: number,
  ): number {
    const R = 6371000;
    const toRad = (d: number) => (d * Math.PI) / 180;
    const dLat = toRad(lat2 - lat1);
    const dLng = toRad(lng2 - lng1);
    const a =
      Math.sin(dLat / 2) ** 2 +
      Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
    return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  }
}
